rfcs/20231002-Compute-type/README.md

```diff
@@ -0,0 +1,165 @@
+
+# 3.a Implicitly apply up-conversion for int8 weights and floating-point activations.
+
+The current library behavior is that the compute type is inferred by
+the weights type. In other words, int8 weights imply using integer
+arithmetic for primitive computation. We hence cannot change that
+behavior to enable weights up-conversion, since it breaks the current
+semantic of the library.
+
+# 3.b. `force_fpmath` knob
+
+This option just adds a boolean `force_fpmath` knob to the
+primitive descriptor constructor. When set to false (default), the
+behavior is the same as today (weights type and `fpmath_mode` dictate
+compute type).  When set to `true`, int8 inputs are up-converted based
+on `fpmath_mode` attribute (so `f32` by default). This option has the
+benefit to be very simple and apply only to primitives that will
+support up-conversion.
+
+The downside of this approach though is that it cannot be extended to
+support down-conversion to int8 data-type.
+
+# 3.c Extend `fpmath_mode` with a more general `math_mode` (recommended)
+Currently, `fpmath_mode` attribute allows the user to specify
+down-conversions of the inputs from f32 to specific datatype.  Here we
+would extend `fpmath_mode`:
+- to apply to non floating point data-types as input, in particular u8/s8.
+- to allow up-conversions from u8/s8 to a particular floating-point
+  type. For example, passing u8 weights, bf16 activations and
+  `math_mode=bf16` to a matmul primitive should convert its weights to
+  bf16 and then compute.
+
+Currently, `fpmath_mode` is ignored for primitives with integral input
+tensors. Adding integral type support hence can be considered either
+as an extension or as a breaking change.  If we consider it breaking,
+this option would require to deprecate `fpmath_mode` in favor of the
+new `math_mode`.
+
+Now to avoid weird conversions and numerical behavior, `math_mode`
+would behave as follow:
+- up-conversion is possible only from u8/s8 to a floating-point
+  datatype (typically f16/bf16)
+- down-conversion is possible only to a "compatible" floating-point
+  datatype (f32->f16, f32->tf32, f32->bf16). To allow for future
+  extension (e.g. down-conversion to integral type), we would have to
+  properly document that the attribute is not ignored in that
+  case, but simply return unimplemented.
+- other configurations should not be ignore like they currently are,
+  but should return unimplemented (e.g. conversions between two types
+  that are the same size like f16<->bf16, or down conversions to
+  int8).
+
+Pros: 
+- very little changes to API.
+- flexible enough to support practical use cases
+- compatible with current implicit down-conversion semantics (the
+existing default `fpmath_mode` would affect only primitive with
+floating-point `math_mode=f32`).
+- can be extended to support integral down-conversion
+(f32/f16->int8). This would require to specify multiple data-types in
+`math_mode` (e.g. `math_mode=u8s8`, `math_mode_s8u8`, `math_mode_s8`,
```

Comment:
For down-converting, what would be the logic of using signed vs unsigned? Is it up to the user w.r.t. the data in the tensor?